src/packages/curl/curl.cc

```diff
@@ -0,0 +1,304 @@
+
+      result->status_message = status_code_message;
+    }
+  }
+
+  unsigned long index;
+  if ((index = header.find(':')) != std::string::npos) {
+    auto header_key = header.substr(0, index);
+    auto header_value = header.substr(index + 1);
+
+    result->headers.insert(std::make_pair(header_key, trim(header_value)));
+  }
+
+  return nitems * size;
+}
+
+curl_slist *headers_mapping_to_curl(mapping_t *headers) {
+  auto *curl_headers = new curl_slist;
+
+  if (headers && headers->table) {
+    mapTraverse(headers, [](mapping_t *, mapping_node_t *node, void *curl_headers_ptr) -> int {
+        auto *curl_headers = static_cast<curl_slist *>(curl_headers_ptr);
+        std::string key = std::string(node->values[0].u.string);
+        std::string value = std::string(node->values[1].u.string);
+        std::string header = key + ": " + value;
+
+        curl_slist_append(curl_headers, header.c_str());
+
+        return 0;
+    }, &curl_headers);
+  }
+
+  return curl_headers;
+}
+
+LPC_INT do_curl(LPC_STRING url, mapping_t *headersArg = nullptr, const std::function<void (CURL *)> &callback = nullptr) {
+  CURLcode res;
+  auto *response_body = new std::string;
+  auto *result = new curl_result_t;
+
+  if (!curl) {
+    if (!(curl = curl_easy_init())) {
+      error("Failed call to curl_easy_init()");
+    }
+
+    curl_results = new std::unordered_map<LPC_INT, curl_result_t *>;
+  }
+
+  curl_easy_setopt(curl, CURLOPT_URL, url);
+  curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, &write_callback);
+  curl_easy_setopt(curl, CURLOPT_WRITEDATA, response_body);
+  curl_easy_setopt(curl, CURLOPT_HEADERFUNCTION, header_callback);
+  curl_easy_setopt(curl, CURLOPT_HEADERDATA, result);
+
+  curl_slist *headers = nullptr;
+
+  if (headersArg) {
+    headers = headers_mapping_to_curl(headersArg);
+    curl_easy_setopt(curl, CURLOPT_HEADER, headers);
+  }
+
+  callback(curl);
+
+  res = curl_easy_perform(curl);
+
+  curl_easy_cleanup(curl);
+  if (headers) {
+    curl_slist_free_all(headers);
+  }
+
+  /* Check for errors */
+  if(res != CURLE_OK) {
+    error("curl_post(%s) failed: %s\n", url, curl_easy_strerror(res));
+  }
+
+  result->body = *response_body;
+
+  auto result_key = get_next_result_id();
+
+  curl_results->insert(std::make_pair(result_key, result));
+
+  return result_key;
+}
+
+#ifdef F_CURL_GET
+void f_curl_get(void) {
+  LPC_STRING url;
+  mapping_t *headers = nullptr;
+  int num_arg = st_num_arg;
+
+  CHECK_TYPES(sp, T_STRING, 0, F_CURL_GET);
+
+  url = (sp - num_arg + 1)->u.string;
+
+  if (num_arg > 1) {
+    CHECK_TYPES(sp, T_MAPPING, 1, F_CURL_GET);
+
+    headers = (sp - num_arg + 2)->u.map;
+  }
+
+  auto result_id = do_curl(url, headers);
+
+  pop_n_elems(num_arg);
+  push_number(result_id);
+}
+#endif
+
+#ifdef F_CURL_POST
+void f_curl_post(void) {
+  LPC_STRING url;
+  LPC_STRING data;
+  mapping_t *headers = nullptr;
+  int num_arg = st_num_arg;
+
+  CHECK_TYPES(sp, T_STRING, 0, F_CURL_POST);
```

Comment:
There is no need to check type for non-mixed argument, the driver automatically enforce the declaration, so you can remove this